src/beamlime/constructors/injection.py

```diff
@@ -0,0 +1,171 @@
+        except ProviderNotFoundError as err:
+            from inspect import Signature
+
+            from .inspectors import UnknownType
+
+            if (arg_spec.default_product != Signature.empty) or (
+                arg_spec.product_type == UnknownType
+            ):
+                return Empty
+            raise err
+
+    def build_arguments(self, provider: Provider):
+        return {
+            arg_name: _product
+            for arg_name, arg_spec in provider.arg_dep_specs.items()
+            if (_product := self.build_arg_dependency(arg_spec)) != Empty
+        }
+
+    def build_attr_dependency(self, dep_type: ProductType):
+        try:
+            return self.assemble(dep_type)
+        except ProviderNotFoundError:
+            return Empty
+
+    def inject_dependencies(self, _obj: _Product) -> _Product:
+        """
+        Check if the ``incomplete_obj`` has attribute dependencies to be filled.
+        If ``incomplete_obj`` is missing any attribute with type hint,
+        or if the attribute is None, the dependency will be assembled and injected.
+        If a provider is not found but the ``incomplete_obj`` already has the attribute,
+        it skips the attribute injection.
+
+        Raises
+        ------
+            ProviderNotFoundError:
+                If a provider is not found for the attribute and
+                the ``incomplete_obj`` still doesn't have the attribute.
+
+        Attribute annotations are retrieved by typing.get_type_hints.
+        """
+        from typing import get_type_hints
+
+        _attrs: Dict[str, Any] = {
+            _attr_name: self.build_attr_dependency(attr_type)
+            for _attr_name, attr_type in get_type_hints(_obj.__class__).items()
+            if not hasattr(_obj, _attr_name) or getattr(_obj, _attr_name) is None
+        }
+
+        if any(
+            missing_attrs := [
+                _attr_name for _attr_name, _attr in _attrs.items() if _attr is Empty
+            ]
+        ):
+            raise ProviderNotFoundError(
+                f"Provider for the attribute(s) {missing_attrs} not found "
+                f"during the dependency injection to the {_obj.__class__} instance."
+            )
+
+        for _attr_name, _attr in _attrs.items():
+            if _attr is Empty:
+                ...  # Skip injecting dependency.
+                # The other case where attr_subprovider
+                # is not found but the ``_obj`` does not
+                # have the attribute is filtered
+                # by checking the missing attributes above.
+            else:
+                setattr(_obj, _attr_name, _attr)
+        return _obj
+
+    def assemble(self, product_type: Type[_Product]) -> _Product:
+        _provider = self.find_provider(product_type)
+        kwargs = {**self.build_arguments(_provider), **_provider.keywords}
+        _obj = _provider.constructor(*_provider.args, **kwargs)
+        return self.inject_dependencies(_obj)
```

Comment:
I'm confused. It looks like after `_obj = _provider.constructor(*_provider.args, **kwargs)`, `_obj` is a `Constructor`. Then, `self.inject_dependencies(_obj)` sets a number of attributes on that `Constructor` but never calls it.
So when does it make a `_Product`?

Reply:
So the ``constructor`` is a property so its return type is ``Constructor`` which is callable, and the return type of the ``constructor`` call is ``_Product``.
I should have called the provider directly instead of calling constructor ... and use type hint in the __call__.
I'm on it thank you...!

Reply:
It is now updated. It might still be confusing that it doesn't allow position-only arguments.
The idea was to annotate all the arguments as non-position-only, but then we might not be able to use external functions from other libraries with this injection framework.
But I don't expect that to happen often anyways so I would rather keep it this way.